In [1]:
import pandas as pd
df = pd.read_csv("usarrests.csv")
df

,Unnamed: 0,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91
5,Colorado,7.9,204.0,78
6,Connecticut,3.3,110.0,77
7,Delaware,5.9,238.0,72
8,Florida,15.4,335.0,80
9,Georgia,17.4,NaN,60


In [2]:
df.dtypes

Unnamed: 0        str
Murder        float64
Assault       float64
UrbanPop        int64
dtype: object

By looking at the table, I can see that the first column is named "Unnamed: 0" while it could be named as "State" or "State Name". I decided to rename the column as "State".

In [3]:
df = df.rename(columns={"Unnamed: 0" : "State"})
df.shape

(50, 4)

In [4]:
df.isna().sum()

State       0
Murder      0
Assault     1
UrbanPop    0
dtype: int64

By running the .isna() and .sum() functions to count the number of na in the table, I was able to count 1 missing data for Georgia's Assault value. I decided to drop the row for Georgia because the assault rate can't possibly be 0 as a typo when its murder rate is 17.4.

In [5]:
df = df.drop(index=9)
df.shape

(49, 4)

In [6]:
df.isna().sum()

State       0
Murder      0
Assault     0
UrbanPop    0
dtype: int64

In [7]:
df.describe()

,Murder,Assault,UrbanPop
count,49.000000,49.000000,49.000000
mean,7.591837,182.183673,74.489796
std,4.171522,130.877435,74.140105
min,0.800000,45.000000,6.000000
25%,4.000000,109.000000,53.000000
50%,7.200000,159.000000,66.000000
75%,11.100000,249.000000,78.000000
max,16.100000,879.000000,570.000000


Based on the describe() function, I see that New York has the minimum urban population with only 6%. This makes zero sense and almost impossible especially when it's compared to other states, so I decided to drop it from the data.

In [8]:
df = df.drop(index=31)
df.shape

(48, 4)

In [9]:
df

,State,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91
5,Colorado,7.9,204.0,78
6,Connecticut,3.3,110.0,77
7,Delaware,5.9,238.0,72
8,Florida,15.4,335.0,80
10,Hawaii,5.3,46.0,83


By looking at the data, I see Iowa's urban population is 570% which is purely impossible. I'm going to treat it as a typo and change it to 57%.

In [10]:
a = df["UrbanPop"][14]
b = 57
df = df.replace(a, b)
df

,State,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91
5,Colorado,7.9,204.0,78
6,Connecticut,3.3,110.0,77
7,Delaware,5.9,238.0,72
8,Florida,15.4,335.0,80
10,Hawaii,5.3,46.0,83


I see that South Carolina's Assault rate at 879.0 seems a little too extreme. I am going to drop the row for South Carolina.

In [11]:
df = df.drop(index=39)
df.shape

(47, 4)

In [12]:
df

,State,Murder,Assault,UrbanPop
0,Alabama,13.2,236.0,58
1,Alaska,10.0,263.0,48
2,Arizona,8.1,294.0,80
3,Arkansas,8.8,190.0,50
4,California,9.0,276.0,91
5,Colorado,7.9,204.0,78
6,Connecticut,3.3,110.0,77
7,Delaware,5.9,238.0,72
8,Florida,15.4,335.0,80
10,Hawaii,5.3,46.0,83


In [13]:
df.describe()

,Murder,Assault,UrbanPop
count,47.000000,47.000000,47.000000
mean,7.372340,165.829787,65.595745
std,4.102899,83.263155,14.377911
min,0.800000,45.000000,32.000000
25%,3.900000,107.500000,55.000000
50%,6.800000,156.000000,66.000000
75%,10.200000,243.500000,77.500000
max,16.100000,337.000000,91.000000


Now, the data seems much more accurate than the previous version.

# Cleaning Log
First, I realized that the first column name with name of the states was just "Unnamed: 0" so I renamed it as "State". Then, I took a look at data and found out the state of Georgia had assault rate of NaN when its murder rate was at 17.4, so I decided to drop the row for Georgia. I also realized New York had only 6% of urban population which seems totally off so I decided to drop the row for New York. Iowa's urban population was at 570% which is impossible and I treated as a typo and switched it to 57%. Finally, I dropped South Carolina's row because its assault rate was at 879.0 which seemed off as well. Overall, 50 loaded; dropped 3 leaving 47.

Repo Address: https://github.com/yeonghwang-hub/cs82a-portfolio